core/src/autogluon/core/callbacks/_abstract_callback.py

```diff
@@ -0,0 +1,205 @@
+        For example, a recursive call can happen if inside `self.before_model_fit` or `self.after_model_fit`, the callback initiates a model fit in trainer.
+        This model fit will then trigger `self.before_model_fit` again, which could lead to an infinite loop if the callback is not implemented carefully.
+        If False, guarantees that the callback logic will be skipped if it is part of a recursive call.
+    skip_if_trainer_stopped : bool, default = False
+        If True, will skip self.before_model_fit and self.after_model_fit logic if `early_stop=True` was returned from any callback,
+        indicating that the trainer is stopping training.
+        This matters if you have 2 or more callbacks, and the first callback returns `early_stop=True`.
+        If the second callback has `skip_if_trainer_stopped=True`, it will skip its callback logic.
+        Otherwise, its callback logic will still trigger.
+
+    Examples
+    --------
+    >>> from autogluon.core.callbacks import ExampleCallback
+    >>> from autogluon.tabular import TabularDataset, TabularPredictor
+    >>> callbacks = [ExampleCallback()]
+    >>> train_data = TabularDataset('https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')
+    >>> label = 'class'
+    >>> predictor = TabularPredictor(label=label).fit(train_data, callbacks=callbacks)
+    """
+
+    allow_recursive_calls: bool = False
+    skip_if_trainer_stopped: bool = False
+
+    def __init__(self):
+        self._skip = False
+
+    def before_trainer_fit(self, trainer: AbstractTrainer, **kwargs):
+        """
+        Called before fitting a trainer.
+        This will be the first method call to the callback at the start of the training process (prior to `before_model_fit`).
+        All available arguments are passed as kwargs along with the trainer object.
+        This allows this method to theoretically override the entire training logic if desired.
+
+        Parameters
+        ----------
+        trainer : AbstractTrainer
+            The AutoGluon trainer object
+        **kwargs
+            Arguments passed to the trainer object's `train_multi_levels` method.
+            Contains all relevant information to completely override the trainer's logic if desired.
+            Refer to the source code for more details, or use a debugger to see the contents of `**kwargs`.
+        """
+        pass
+
+    def after_trainer_fit(self, trainer: AbstractTrainer):
+        """
+        Called after fitting a trainer.
+        This will be the final method call to the callback before AutoGluon training completes.
+        Example usages of this method include logging a final summary of the training, saving information to disk, or executing additional post-fit logic.
+        """
+        pass
+
+    def before_model_fit(
+        self,
+        trainer: AbstractTrainer,
+        model: AbstractModel,
+        time_limit: float | None = None,
+        stack_name: str = "core",
```

Comment:
Should this be a `typing.Literal`?

Reply:
in this case no because stack_name can be any valid string.